src/serac/infrastructure/set.hpp

```diff
@@ -0,0 +1,324 @@
+    Set<T> diff;
+    for (auto s1_key : keys) {
+      auto s1_diff = Difference(values_index_list[s1_key], s2.values());
+      if (diff.values_index_list.find(s1_key) == diff.values_index_list.end()) {
+        // key doesn't exist yet
+        diff.keys.push_back(s1_key);
+        diff.values_index_list[s1_key] = s1_diff;
+      } else {
+        // append to previous values
+        diff.values_index_list[s1_key] = Union(diff.values_index_list[s1_key], s1_diff);
+      }
+    }
+
+    // Sort the keys and sort each of the values for each key
+    std::sort(diff.keys.begin(), diff.keys.end());
+    for (auto& [key, list] : diff.values_index_list) {
+      std::sort(list.begin(), list.end());
+      diff.total_size += list.size();
+    }
+
+    diff.removeEmptyValues();
+    return diff;
+  }
+
+  /// Get the complement of a given subset of keys
+  Set<T> getComplement(std::vector<T> subkeys)
+  {
+    // find the difference of the keys
+    Set<T> sall;
+    sall.keys = Difference(keys, subkeys);
+    // copy over only the vectors that are part of the commplement
+    for (auto k : sall.keys) {
+      sall.values_index_list[k] = values_index_list[k];
+      sall.total_size += sall.values_index_list[k].size();
+    }
+    return sall;
+  }
+
+  /// Get the subset of a given subset of keys
+  Set<T> getSubset(std::vector<T> subkeys)
+  {
+    // find the difference of the keys
+    Set<T> sall;
+    sall.keys = subkeys;
+    // copy over only the vectors that are part of the subset
+    for (auto k : sall.keys) {
+      sall.values_index_list[k] = values_index_list[k];
+      sall.total_size += sall.values_index_list[k].size();
+    }
+    return sall;
+  }
+
+  /// Remove empty values and keys
+  void removeEmptyValues()
+  {
+    // some values_index_lists[key] might be empty. If so get rid of them
+    for (auto k = keys.begin(); k != keys.end();) {
+      if (values_index_list[*k].size() == 0) {
+        k = keys.erase(k);
+      } else {
+        k++;
+      }
+    }
+
+    for (auto it = values_index_list.begin(); it != values_index_list.end();) {
+      if (values_index_list[it->first].size() == 0) {
+        it = values_index_list.erase(it);
+      } else {
+        it++;
+      }
+    }
+  }
+
+  /// Allows us to print sets to stream
+  friend std::ostream& operator<<(std::ostream& os, const Set<T>& set)
+  {
+    for (auto k : set.keys) {
+      os << k << " : ";
+      for (auto i : set.values_index_list.at(k)) {
+        os << i << " ";
+      }
+      os << std::endl;
+    }
+    return os;
+  }
+
+  /*
+   * @brief  Method to check if two sets are the same
+   * @param[in] set1
+   * @param[in[ set2
+   */
+  friend bool operator==(const Set<T>& s1, const Set<T>& s2)
+  {
+    bool isSame = (s1.keys == s2.keys);
+    if (isSame) {
+      for (auto k : s1.keys) {
+        isSame = isSame && (s1.values_index_list.at(k) == s2.values_index_list.at(k));
+        if (!isSame) break;
+      }
+    }
+    return isSame;
```

Comment:
`isSame` -> `is_same`

Minor: the control flow can probably be cleaned up a little bit since I don't think we're worried about multiple returns